mindsdb/integrations/handlers/anomaly_detection_handler/anomaly_detection_handler.py

```diff
@@ -0,0 +1,172 @@
+    if model_name is not None:
+        assert model_name in MODELS[model_type], f"model name must be one of {list(MODELS[model_type].keys())}"
+        model = MODELS[model_type][model_name]
+    else:
+        model = None
+    if model_type == "unsupervised":
+        return train_unsupervised(training_df, model=model)
+
+    X_train = training_df.drop(target, axis=1)
+    y_train = training_df[target].astype(int)
+
+    if model_type == "supervised":
+        return train_supervised(X_train, y_train, model=model)
+    elif model_type == "semi-supervised":
+        return train_semisupervised(X_train, y_train)  # Only one semi-supervised model available
+
+
+def anomaly_type_to_model_name(anomaly_type):
+    """Choose the best model name based on the anomaly type"""
+    assert anomaly_type in [
+        "local",
+        "global",
+        "clustered",
+        "dependency",
+    ], "anomaly type must be 'local', 'global', 'clustered', or 'dependency'"
+    anomaly_type_dict = {
+        "local": "lof",
+        "global": "knn",
+        "clustered": "pca",
+        "dependency": "knn",
+    }
+    return anomaly_type_dict[anomaly_type]
+
+
+def preprocess_data(df):
+    """Preprocess the data by one-hot encoding categorical columns and scaling numeric columns"""
+    # one-hot encode categorical columns
+    categorical_columns = list(df.select_dtypes(include=["object"]).columns.values)
+    df[categorical_columns] = df[categorical_columns].astype("category")
+    df[categorical_columns] = df[categorical_columns].apply(lambda x: x.cat.codes)
+    df = pd.get_dummies(df, columns=categorical_columns)
+    numeric_columns = list(df.select_dtypes(include=["number"]).columns.values)
+    df[numeric_columns] = (df[numeric_columns] - df[numeric_columns].mean()) / df[numeric_columns].std()
+    return df
+
+
+def get_model_names(using_args):
+    """Get the model names from the using_args. Model names is a list of model names to train.
+    If the model is not an ensemble, it only contains one model"""
+    model_names = anomaly_type_to_model_name(using_args["anomaly_type"]) if "anomaly_type" in using_args else None
+    model_names = using_args["model_name"] if "model_name" in using_args else model_names
+    model_names = using_args["ensemble_models"] if "ensemble_models" in using_args else model_names
+    model_names = [model_names] if model_names is None else model_names
+    model_names = [model_names] if type(model_names) is str else model_names
+    return model_names
+
+
+class AnomalyDetectionHandler(BaseMLEngine):
+    """Integration with the PyOD and CatBoost libraries for
+    anomaly detection. Both supervised and unsupervised.
+    """
+
+    name = "anomaly_detection"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.generative = True  # makes unsupervised learning work
+
+    def create(self, target, df, args={}):
```

Comment:
{} is mutable default value. 
here should be 'args=None'

the same is for .predict method

Reply:
Didn't realise this - thanks for point it out